README.md
# Curity Cypress Commands

Cypress commands used when testing the Curity Identity Server plugins.

## Installation

Install using npm, with

```shell
npm install -D @curity/cypress-commands
```

## Usage

To use the commands in your tests:

1. Import the registration function in your spec file with `import { registerCurityCommands } from '@curity/cypress-commands`.
2. Register commands once before running any tests by calling `registerCurityCommands()`.
3. You can then use the commands as any other Cypress command, e.g. `cy.getIDToken()`.

You can see the spec files in `cypress/integration` to check how the commands are used in Cypress tests. 

### Registering with a Prefix

If you happen to use other custom Cypress commands with the same names, then you can register the Curity commands with a prefix.
E.g. `registerCurityCommands('curity')`. All command names from this package will be prefixed with the chosen string and the
first letter of the original command name will be capitalized. E.g. `getIDToken` will become `curityGetIDToken`.

Note that you will not get Typescript support if you register commands with a prefix.

## Available commands

- [buildAuthorizationURL](#buildauthorizationurl)
- [startAuthorization](#startauthorization)
- [decodeJWT](#decodejwt)
- [getIDToken](#getidtoken)
- [getIDTokenClaims](#getidtokenclaims)

### buildAuthorizationURL

Prepares an authorization URL object based on the passed parameters. See `src/index.d.ts` for the list of currently supported parameters.

Example usage

```javascript
    cy.buildAuthorizationURL({baseURL: "https://idsvr.example.com", clientID: "client1"})
        .then(url => cy.visit(url.toString())).get('#login').should('exist')
```

### startAuthorization

Visits the Authorization URL. Either the previous subject needs to be a URL or the parameters object must be passed.
See `src/index.d.ts` for the list of currently supported parameters.

Example usage

```javascript
    const parameters = { baseURL: "https://idsvr.example.com", clientID: "client1" }
    cy.buildAuthorizationURL(parameters).startAuthorization().get('#login').should('exist')
    cy.startAuthorization(parameters).get('#login').should('exist')
```

### decodeJWT         

Decodes a JWT and returns its claims in form of a map. If no parameter is passed then the previous subject is taken as the token.

Example usage

```javascript
    cy.decodeJWT(token).its('sub').should('equal', 'john.doe')
```

### getIDToken

Gets the ID token from the fragment part of the current URL.

Example usage

```javascript
    cy.getIDToken().should('not.be.empty')
```

### getIDTokenClaims

Gets the ID token from the fragment part of the current URL, decodes it and returns its claims. It's a shorthand for calling
`getIDToken` and `decodeJWT` in a chain.

Example usage

```javascript
    cy.getIDTokenClaims().its('sub').should('equal', 'john.doe')
```


CHANGELOG.md
# Curity Cypress Commands Changelog

## Unreleased

- Fix: minor change in handling of map of extra params.

## 1.0.0 (2022-06-07)

Initial release of the commands library.


cypress/integration/curity-commands.spec.js
/*
 * Copyright 2022 Curity AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/// <reference path="../../src/index.d.ts" />

import { registerCurityCommands } from "../../src";

describe('Curity Cypress commands tests', () => {
  const idToken = 'eyJhbGciOiJSUzI1NiJ9.eyJzdWIiOiJqb2huLmRvZSIsImlzcyI6Imh0dHBzOi8vaWRzdnIuZXhhbXBsZS5jb20iLCJhdWQiOiJ0ZXN0LWNsaWVudCJ9.VaG1QyN_etsHSFayUspDJ8wxPCEAr3bLz6hNExA87KcMurAbOWmmsIc8lTFxBNitPV7gov-aHf85vTiPUbefY8uQRF0BdNKFamCpznWQRbdi06xmyr5Pc6GUftarfZYackklU6W27qNWG6An05faLpJV887Hp4WDONQSFOr3_3UB9lU3XNR3mFKdteqccHTm4hc_9JDsExeEZNN52izOKhYJFc_ZKYr52b16KLsJ4qYrnCOzEWRmYRr1iKIxRWjHa_lc4pPYeg3958CewBaz_8Y1QPSXgpPga8o4gGt4MEs_NL6kqkjGv24SBSxpFtR8h4JsiTncL4YPiQNj2isQT_XDTC92DCJo_b4Y5QUxAgxsvEualBQ7zZ0iqvUrjyl9c2udd-QsI9IyuOW5l2jhVBo9eI4qsyqo2B_7mU8PNzgOTFw-0n_qA5zrF2OHNqUXKPNGgk5JSG-AOUfex3E9eMcTaHAjhy1kWeanWF-9BBcTsgp9UMIShzBvnx39JigO_iS-azYaR-jXccEsTHb56EATwTFBKBpSq9HnUSEZoEDMpnME2k_7ANohgxJFIAOsEiaf9RDwV7HMwZxEVgxbt9FOirhDc1vATSmGzq0ia5zSnftTTFWAwoCuQekRf7HsroGvCO8dO-Ad6HT1KKIZQDpg_1IXe8BoCkNGs_GXLtk'

  registerCurityCommands()

  it('Default parameters are added to the authorization URL', () => {

    cy.buildAuthorizationURL({
      baseURL: 'https://idsvr.example.com/authorize',
      clientID: 'test-client'
    })
    .then(url => {
      expect(url.toString())
          .to.match(new RegExp('^https://idsvr\\.example\\.com/authorize\\?'))
          .and
          .contains('response_type=code')
          .and
          .contains('prompt=login')
    })
  })

  it('All passed parameters are added to the authorization URL', () => {

    cy.buildAuthorizationURL({
      baseURL: 'https://idsvr.example.com/authorize',
      clientID: 'test-client',
      responseType: 'code id_token',
      redirectURI: 'http://localhost/cb',
      state: 'abc',
      prompt: 'none',
      extraParams: {
        code_challenge: '1234',
        code_challenge_method: 'plain'
      }
    })
        .then(url => {
          expect(url.toString())
              .to.match(new RegExp('^https://idsvr\\.example\\.com/authorize\\?'))
              .and.contain('response_type=code+id_token')
              .and.contain('prompt=none')
              .and.contain('client_id=test-client')
              .and.contain('redirect_uri=http%3A%2F%2Flocalhost%2Fcb')
              .and.contain('state=abc')
              .and.contain('code_challenge=1234')
              .and.contain('code_challenge_method=plain')
        })
  })

  it('Should start authorization when parameters are passed', () => {
    const parameters = { baseURL: "https://idsvr.example.com/authorize", clientID: "test-client" }
    let requestCounter = 0

    cy.intercept(
        { url: /^https:\/\/idsvr\.example\.com\/authorize/ },
        (req) => {
          requestCounter += 1
          req.reply({
            statusCode: 200,
            body: '',
            headers: {
              'content-type': 'text/html'
            }
          })
    }).as('authorize')

    cy.startAuthorization(parameters)

    cy.wait('@authorize').then(() => {
      expect(requestCounter).to.equal(1)
    })
  })

  it('Should start authorization when no parameters passed but call chained with a URL subject', () => {
    const parameters = { baseURL: "https://idsvr.example.com/authorize", clientID: "test-client" }
    let requestCounter = 0

    cy.intercept(
        { url: /^https:\/\/idsvr\.example\.com\/authorize/ },
        (req) => {
          requestCounter += 1
          req.reply({
            statusCode: 200,
            body: '',
            headers: {
              'content-type': 'text/html'
            }
          })
        }).as('authorize')

    cy.buildAuthorizationURL(parameters).startAuthorization()

    cy.wait('@authorize').then(() => {
      expect(requestCounter).to.equal(1)
    })
  })

  it('Should decode a JWT', () => {
    cy.decodeJWT(idToken).then(claims => {
      expect(claims.sub).to.exist.and.equal('john.doe')
      expect(claims.aud).to.exist.and.equal('test-client')
    })
  })

  it('Should read an ID token from fragment part of current URL', () => {
    cy.visit('cypress/index.html#id_token=abcdef').getIDToken().should('equal', 'abcdef')
  })

  it('Should decode an ID token from fragment part of URL', () => {
    cy.visit(`cypress/index.html#id_token=${idToken}`).getIDTokenClaims().its('sub').should('equal', 'john.doe')
  })

  it('Should decode an ID token from fragment part of URL with custom commands prefix', () => {
    registerCurityCommands("custom")
    cy.visit(`cypress/index.html#id_token=${idToken}`).customGetIDTokenClaims().its('sub').should('equal', 'john.doe')
  })
})


src/index.js
/*
 *  Copyright 2022 Curity AB
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

/// <reference types="Cypress" />


const buildAuthorizationURLCommand = (parameters) => {
    const authorizationURL = new URL(parameters.baseURL)
    const params = authorizationURL.searchParams

    params.append('client_id', parameters.clientID)
    params.append('response_type', parameters.responseType ? parameters.responseType : 'code')
    params.append('prompt', parameters.prompt ? parameters.prompt : 'login')

    if (parameters.redirectURI) {
        params.append('redirect_uri', parameters.redirectURI)
    }

    if (parameters.scope) {
        params.append('scope', parameters.scope)
    }

    if (parameters.state) {
        params.append('state', parameters.state)
    }

    if (parameters.extraParams) {
        for (const [parameterName, parameterValue] of parameters.extraParams) {
            params.append(parameterName, parameterValue)
        }
    }

    return authorizationURL
}

const createStartAuthorizationCommand = (commandPrefix) => {
    const buildAuthorizationURLCommandName = getCommandName(commandPrefix, 'buildAuthorizationURL')
    const startAuthorizationCommandName = getCommandName(commandPrefix, 'startAuthorization')

    return (subject, parameters) => {
        if (subject && subject instanceof URL) {
            cy.visit(subject.toString())
        } else {
            cy[buildAuthorizationURLCommandName](parameters)[startAuthorizationCommandName]()
        }
    }
}

const decodeJWTCommand = (subject, jwt) => {
    const token = (subject !== undefined && typeof subject === 'string') ? subject : jwt
    return JSON.parse(decodeJWT(token))
}

const getIDTokenCommand = () => {
    console.trace()
    cy.hash()
        .should('contain', '#')
        .should('contain', 'id_token')
        .then(url => {
            const fragment = url.split('#')[1]
            let idToken = null

            for (const parameter of fragment.split('&')) {
                if (parameter.startsWith('id_token')) {
                    idToken = parameter.split('=')[1]
                    break
                }
            }

            expect(idToken).not.to.be.empty

            return idToken
        })
}

const createGetIDTokenClaimsCommand = (commandPrefix) => {
    const getIDTokenName = getCommandName(commandPrefix, 'getIDToken')
    const decodeJWTName = getCommandName(commandPrefix, 'decodeJWT')

    return () => {
        cy[getIDTokenName]()[decodeJWTName]()
    }
}

const decodeJWT = (token) => {
    const payload = token.split('.')[1]
    expect(payload).not.to.be.empty

    return decodeURIComponent(Array.prototype.map.call(atob(payload.replace('-', '+').replace('_', '/')), c =>
        '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2)
    ).join(''))
}

const registerCurityCommands = (commandPrefix) => {
    const commands = {
        buildAuthorizationURL: { command: buildAuthorizationURLCommand },
        startAuthorization: { command: createStartAuthorizationCommand(commandPrefix), options: { prevSubject: 'optional' }},
        decodeJWT: { command: decodeJWTCommand, options: { prevSubject: 'optional' }},
        getIDToken: { command: getIDTokenCommand },
        getIDTokenClaims: { command: createGetIDTokenClaimsCommand(commandPrefix) }
    }

    for (const commandName in commands) {
        const command = commands[commandName]
        if (command.options) {
            Cypress.Commands.add(getCommandName(commandPrefix, commandName), command.options, command.command)
        } else {
            Cypress.Commands.add(getCommandName(commandPrefix, commandName), command.command)
        }
    }
}

const getCommandName = (prefix, name) => {
    if (prefix) {
        return prefix + name.charAt(0).toUpperCase() + name.slice(1)
    } else {
        return name
    }
}

export { registerCurityCommands }


src/index.d.ts
/*
 *  Copyright 2022 Curity AB
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

/// <reference types="cypress" />
declare global {
    namespace Cypress {
        interface Chainable {
            /**
             * Prepares an authorization URL object based on the passed parameters
             * @param parameters AuthorizationURLParameters
             * @example
             *  cy.buildAuthorizationURL({baseURL: "...", clientID: "..."}).then(url => cy.visit(url.toString())).get('#login').should('exist')
             */
            buildAuthorizationURL(parameters: AuthorizationURLParameters): Chainable<URL>

            /**
             * Visits the Authorization URL. Either the previous subject needs to be a URL or the parameters object must be passed.
             * @example
             *   - cy.buildAuthorizationURL({baseURL: "...", clientID: "..."}).startAuthorization().get('#login').should('exist')
             *   - cy.startAuthorization({baseURL: "...", clientID: "..."}).get('#login').should('exist')
             */
            startAuthorization(parameters?: AuthorizationURLParameters): Chainable<AUTWindow>

            /**
             * Decodes a JWT and returns its claims in form of an object. If no parameter is passed then the previous subject is taken as the token.
             * @param jwt string
             * @example
             *  cy.decodeJWT(token).its('sub').should('equal', 'john.doe')
             */
            decodeJWT(jwt?: string): Chainable<Map<string, object>>

            /**
             * Gets the ID token from the fragment part of the current URL.
             * @example
             *  cy.getIDToken().should('not.be.empty')
             */
            getIDToken(): Chainable<string>

            /**
             * Gets an ID token from the fragment part of the current URL, decodes it and returns its claims.
             * @example
             *  cy.getIDTokenClaims().its('sub').should('equal', 'john.doe')
             */
            getIDTokenClaims(): Chainable<Map<string, object>>
        }

        interface AuthorizationURLParameters {
            /** The URL of the authorization endpoint. **/
            baseURL: string,
            /** The value of the `client_id` parameter. **/
            clientID: string,
            /** The value of the `response_type` parameter. Defaults to `code`. **/
            responseType?: string,
            /** The value of the `redirect_uri` parameter. **/
            redirectURI?: string,
            /** The value of the `scope` parameter. **/
            scope?: string,
            /** The value of the `state` parameter. **/
            state?: string,
            /** The value of the `prompt` parameter. Defaults to `login` in order to force login for every test (avoid SSO) **/
            prompt?: string,
            /** Any extra parameters can be added as a map in the `extraParams` fields. They will all be added to the resulting URL. **/
            extraParams?: Map<string, string>
        }
    }

}

/**
 * Registers the Curity commands in Cypress.
 *
 * @param commandPrefix Prefix all command names. The resulting command name will have its first letter capitalized. E.g. `getIDToken` becomes `prefixGetIDToken`.
 */
declare function registerCurityCommands(commandPrefix?: string): void

export { registerCurityCommands }


cypress/e2e/curity-commands.cy.js
/*
 * Copyright 2022 Curity AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/// <reference path="../../src/index.d.ts" />


import { registerCurityCommands } from "../../src";

describe('Curity Cypress commands tests', () => {
  const idToken = 'eyJhbGciOiJSUzI1NiJ9.eyJzdWIiOiJqb2huLmRvZSIsImlzcyI6Imh0dHBzOi8vaWRzdnIuZXhhbXBsZS5jb20iLCJhdWQiOiJ0ZXN0LWNsaWVudCJ9.VaG1QyN_etsHSFayUspDJ8wxPCEAr3bLz6hNExA87KcMurAbOWmmsIc8lTFxBNitPV7gov-aHf85vTiPUbefY8uQRF0BdNKFamCpznWQRbdi06xmyr5Pc6GUftarfZYackklU6W27qNWG6An05faLpJV887Hp4WDONQSFOr3_3UB9lU3XNR3mFKdteqccHTm4hc_9JDsExeEZNN52izOKhYJFc_ZKYr52b16KLsJ4qYrnCOzEWRmYRr1iKIxRWjHa_lc4pPYeg3958CewBaz_8Y1QPSXgpPga8o4gGt4MEs_NL6kqkjGv24SBSxpFtR8h4JsiTncL4YPiQNj2isQT_XDTC92DCJo_b4Y5QUxAgxsvEualBQ7zZ0iqvUrjyl9c2udd-QsI9IyuOW5l2jhVBo9eI4qsyqo2B_7mU8PNzgOTFw-0n_qA5zrF2OHNqUXKPNGgk5JSG-AOUfex3E9eMcTaHAjhy1kWeanWF-9BBcTsgp9UMIShzBvnx39JigO_iS-azYaR-jXccEsTHb56EATwTFBKBpSq9HnUSEZoEDMpnME2k_7ANohgxJFIAOsEiaf9RDwV7HMwZxEVgxbt9FOirhDc1vATSmGzq0ia5zSnftTTFWAwoCuQekRf7HsroGvCO8dO-Ad6HT1KKIZQDpg_1IXe8BoCkNGs_GXLtk'

  registerCurityCommands()

  it('Default parameters are added to the authorization URL', () => {

    cy.buildAuthorizationURL({
      baseURL: 'https://idsvr.example.com/authorize',
      clientID: 'test-client'
    })
    .then(url => {
      expect(url.toString())
          .to.match(new RegExp('^https://idsvr\\.example\\.com/authorize\\?'))
          .and
          .contains('response_type=code')
          .and
          .contains('prompt=login')
    })
  })

  it('All passed parameters are added to the authorization URL', () => {

    cy.buildAuthorizationURL({
      baseURL: 'https://idsvr.example.com/authorize',
      clientID: 'test-client',
      responseType: 'code id_token',
      redirectURI: 'http://localhost/cb',
      state: 'abc',
      prompt: 'none',
      extraParams: new Map([
        ['code_challenge', '1234'],
        ['code_challenge_method', 'plain']
  ])
    })
        .then(url => {
          expect(url.toString())
              .to.match(new RegExp('^https://idsvr\\.example\\.com/authorize\\?'))
              .and.contain('response_type=code+id_token')
              .and.contain('prompt=none')
              .and.contain('client_id=test-client')
              .and.contain('redirect_uri=http%3A%2F%2Flocalhost%2Fcb')
              .and.contain('state=abc')
              .and.contain('code_challenge=1234')
              .and.contain('code_challenge_method=plain')
        })
  })

  it('Should start authorization when parameters are passed', () => {
    const parameters = { baseURL: "https://idsvr.example.com/authorize", clientID: "test-client" }
    let requestCounter = 0

    cy.intercept(
        { url: /^https:\/\/idsvr\.example\.com\/authorize/ },
        (req) => {
          requestCounter += 1
          req.reply({
            statusCode: 200,
            body: '',
            headers: {
              'content-type': 'text/html'
            }
          })
    }).as('authorize')

    cy.startAuthorization(parameters)

    cy.wait('@authorize').then(() => {
      expect(requestCounter).to.equal(1)
    })
  })

  it('Should start authorization when no parameters passed but call chained with a URL subject', () => {
    const parameters = { baseURL: "https://idsvr.example.com/authorize", clientID: "test-client" }
    let requestCounter = 0

    cy.intercept(
        { url: /^https:\/\/idsvr\.example\.com\/authorize/ },
        (req) => {
          requestCounter += 1
          req.reply({
            statusCode: 200,
            body: '',
            headers: {
              'content-type': 'text/html'
            }
          })
        }).as('authorize')

    cy.buildAuthorizationURL(parameters).startAuthorization()

    cy.wait('@authorize').then(() => {
      expect(requestCounter).to.equal(1)
    })
  })

  it('Should decode a JWT', () => {
    cy.decodeJWT(idToken).then(claims => {
      expect(claims.sub).to.exist.and.equal('john.doe')
      expect(claims.aud).to.exist.and.equal('test-client')
    })
  })

  it('Should read an ID token from fragment part of current URL', () => {
    cy.visit('cypress/index.html#id_token=abcdef').getIDToken().should('equal', 'abcdef')
  })

  it('Should decode an ID token from fragment part of URL', () => {
    cy.visit(`cypress/index.html#id_token=${idToken}`).getIDTokenClaims().its('sub').should('equal', 'john.doe')
  })

  it('Should decode an ID token from fragment part of URL with custom commands prefix', () => {
    registerCurityCommands("custom")
    cy.visit(`cypress/index.html#id_token=${idToken}`).customGetIDTokenClaims().its('sub').should('equal', 'john.doe')
  })
})
